ecleankernel/layout/blspec.py

```diff
@@ -29,7 +29,7 @@ class BlSpecLayout(ModuleDirLayout):
     """
 
     name = 'blspec'
-    potential_dirs = ('boot/EFI', 'boot/efi', 'boot', 'efi')
+    potential_dirs = ('boot/EFI', 'boot/EFI/EFI', 'boot/EFI/efi', 'boot/efi', 'boot/efi/EFI', 'boot/efi/efi', 'efi')
```

Comment:
I'm not good at combinatorics but this starts to look like it'd use one of them functions.

Reply:
I'll look into making this a bit more clever later today.

Reply:
Fixed using the `product()` function. 

Takes a couple more lines of code though, the output now is:
```
('efi/efi/', 'boot/efi/efi/', 'efi/EFI/', 'boot/efi/EFI/', 'efi//', 'boot/efi//', 'EFI/efi/', 'boot/EFI/efi/', 'EFI/EFI/', 'boot/EFI/EFI/', 'EFI//', 'boot/EFI//')
```

Reply:
There must be a better way…

Reply:
`90-uki-copy.install` copies the UKI's to `UKI_DIR="$BOOT_ROOT/EFI/Linux"` so we _might_ be able to get away with simply using `/boot/EFI` only.

Reply:
Cause I've seen configurations where ESP is mounted at /boot/efi instead of /boot, that would make the full path of the directory containing the uki's /boot/efi/EFI/Linux/.

But now that I think about it, the final part always has to be EFI, so probably we can restrict the options to:
- /boot/EFI (ESP mounted at /boot) 
- /boot/efi/EFI, /boot/EFI/EFI (ESP mounted in /boot/efi or /boot/EFI)
- /efi/EFI, /EFI/EFI (ESP mounted at /efi or /EFI) Do people actually do this? The existing code seems to think so, but this I have never seen in practice.

Reply:
But if the final `/EFI` is part of uki layout, why do you also apply it to the regular layout? Wouldn't it make more sense to leave the old list in, and just append `/EFI` for uki check?

Reply:
It's not part of the regular layout, the regular layout is always /root/boot/ENTRY_TOKEN.

I don't know why the code as it was looks for type 1 boot files in ESP, this is clearly incorrect according to the specification. I assumed this was an (incorrect) attempt at implementing the type 2 spec.

Reply:
No. It is because ESP is sometimes mounted at `/boot` and sometimes at `/boot/efi` (or `EFI`, sigh).

Reply:
Yes but type 1 files are not installed to the ESP (unless the ESP is at /boot)

Reply:
Also there is an edge case where ENTRY_TOKEN=linux or Linux, if we check for type 1 files in /boot/EFI this will trigger a false positive and create a mess, potentially removing UKI files that shouldn't be removed. This is why I only check for type 1 files in /boot/ and not in /boot/EFI or /boot/efi/EFI.

Any ENTRY_TOKEN that happens to also be a directory in $ESP/EFI will trigger this problem. So lets just avoid this and only check for type 1 files in $BOOT/$ENTRY_TOKEN and type 2 files in $ESP/EFI/Linux. We still need the `potential_dirs` logic to find the ESP, but it is not used for the type 1 files.

Reply:
> There must be a better way…

`bootctl status` can tell us where ESP and BOOT are. Would this be a viable solution? It will probably involve some magic to properly parse the output and extract the information we want.

Reply:
Fixed! 

This solution avoids guessing where the boot entries are and thus avoids breaking things in these edge cases I mentioned.